src/main/java/net/fabricmc/lorenztiny/TinyMappingsJoiner.java

```diff
@@ -0,0 +1,120 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package net.fabricmc.lorenztiny;
+
+import net.fabricmc.mapping.tree.ClassDef;
+import net.fabricmc.mapping.tree.FieldDef;
+import net.fabricmc.mapping.tree.MethodDef;
+import net.fabricmc.mapping.tree.TinyTree;
+import org.cadixdev.lorenz.MappingSet;
+import org.cadixdev.lorenz.io.MappingsReader;
+import org.cadixdev.lorenz.model.ClassMapping;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A {@link MappingsReader mappings reader} for reading mappings
+ * from across two {@link TinyTree tiny trees}, using a namespace
+ * present in both to match.
+ *
+ * @author Jamie Mansfield
+ * @since 2.0.0
+ */
+public class TinyMappingsJoiner extends MappingsReader {
+
+	private final TinyTree treeA;
+	private final String from;
+	private final String matchA;
+
+	private final TinyTree treeB;
+	private final String to;
+	private final String matchB;
+
+	public TinyMappingsJoiner(final TinyTree treeA, final String from, final String matchA,
+	                          final TinyTree treeB, final String to, final String matchB) {
+		this.treeA = treeA;
+		this.from = from;
+		this.matchA = matchA;
+		this.treeB = treeB;
+		this.to = to;
+		this.matchB = matchB;
+	}
+
+	public TinyMappingsJoiner(final TinyTree treeA, final String from,
+	                          final TinyTree treeB, final String to,
+	                          final String match) {
+		this(treeA, from, match, treeB, to, match);
+	}
+
+	@Override
+	public MappingSet read(final MappingSet mappings) throws IOException {
+		// These maps have matched name -> definition from a
+		final Map<String, ClassDef> classes = new HashMap<>();
+		final Map<String, FieldDef> fields = new HashMap<>();
+		final Map<String, MethodDef> methods = new HashMap<>();
+
+		for (final ClassDef klass : this.treeA.getClasses()) {
+			classes.put(klass.getName(this.matchA), klass);
+
+			for (final FieldDef field : klass.getFields()) {
+				fields.put(field.getName(this.matchA), field);
+			}
+
+			for (final MethodDef method : klass.getMethods()) {
+				methods.put(method.getName(this.matchA), method);
+			}
+		}
+
+		for (final ClassDef classB : this.treeB.getClasses()) {
+			final ClassDef classA = classes.get(classB.getName(this.matchB));
+
+			final ClassMapping<?, ?> mapping = mappings.getOrCreateClassMapping(classA.getName(this.from))
+					.setDeobfuscatedName(classB.getName(this.to));
+
+			for (final FieldDef fieldB : classB.getFields()) {
+				final FieldDef fieldA = fields.get(fieldB.getName(this.matchB));
+
+				mapping.getOrCreateFieldMapping(fieldA.getName(this.from), fieldA.getDescriptor(this.from))
```

Comment:
Im getting an NPE here:

```
Caused by: java.lang.NullPointerException: Cannot invoke "net.fabricmc.mapping.tree.FieldDef.getName(String)" because "fieldA" is null
	at net.fabricmc.lorenztiny.TinyMappingsJoiner.read(TinyMappingsJoiner.java:101)
	at org.cadixdev.lorenz.io.MappingsReader.read(MappingsReader.java:53)
	at net.fabricmc.loom.task.MigrateMappingsTask.migrateMappings(MigrateMappingsTask.java:153)
	at net.fabricmc.loom.task.MigrateMappingsTask.doTask(MigrateMappingsTask.java:100)
```

Seems to happen when the target mappigns have a field that the source mappings dont?

In loom I have:

```java
MappingSet mappingSet = new TinyMappingsJoiner(currentMappings, "named", targetMappings, "named", "intermediary").read();
```

The old code was:

```java
MappingSet mappingSet = LorenzTiny.readMappings(currentMappings, targetMappings, "intermediary", "named").read();
```